crypto/digestible/src/lib.rs

```diff
@@ -3,28 +3,195 @@
+    #[inline]
+    fn digest32<DT: DigestTranscript>(&self, context: &'static [u8]) -> [u8; 32] {
+        let mut transcript = DT::new();
+        self.append_to_transcript(context, &mut transcript);
+        let mut result = [0u8; 32];
+        transcript.extract_digest(&mut result);
+        result
+    }
+
+    /// To support schema evolution, in some contexts the generated code for a Digestible
+    /// implementation of a compound type should append its children to the transcript,
+    /// but allow them to skip themselves if they are empty.
+    /// For members of a struct, this is allowed, because they don't have a fixed set of members
+    /// but for a variant it isn't allowed, because the value cannot be omitted.
+    ///
+    /// The user of the library should normally not call this directly, or override it.
+    /// It has special implementations for types like Option.
+    #[inline]
+    fn append_to_transcript_allow_omit<DT: DigestTranscript>(
         &self,
-    ) -> generic_array::GenericArray<u8, <D as Digest>::OutputSize> {
-        let mut hasher = D::default();
-        self.digest(&mut hasher);
-        hasher.finalize()
+        context: &'static [u8],
+        transcript: &mut DT,
+    ) {
+        self.append_to_transcript(context, transcript)
+    }
+}
+
+/// A trait implemented by protocol transcript objects.
+/// This represents the functions of merlin::Transcript that we need for digestible
+/// to function. By having a trait, we can easily substitute mock objects for tests,
+/// etc.,
+pub trait DigestTranscript {
+    // These low-level calls are needed to implement digest32, digest64 etc.,
+    // and correspond to raw calls on a MerlinTranscript
+    fn new() -> Self;
+    fn append_bytes(&mut self, context: &'static [u8], data: impl AsRef<[u8]>);
+    fn extract_digest(self, output: &mut [u8; 32]);
+
+    // These high-level calls should be used exclusivley when implementing digestible.
+    // These four calls correspond to the four types of nodes in the AST
+    // that is being encoded into the digest.
+    //
+    // Not recommended to override any of these, except for testing purposes
+
+    /// Append a primitive with particular context string, type name, and byte data.
+    ///
+    /// The context string comes from the caller context.
+    /// If the primitive is a struct member, this is its field name.
+    /// If the primitive is an enum member, this is the variant name.
+    /// If the primitive is the only thing being digested, the context string
+    /// comes from the call to digest32 / digest64.
+    ///
+    /// The typename is an arbitrary string representing the type of the data,
+    /// for example, `uint`, `bytes`, `string`, `ristretto`
+    ///
+    /// The data is the canonical bytes representing the primitive.
+    /// If the primitive does not have a canonical representation as bytes then
+    /// it isn't appropriate to treat it as a primitive in this hashing scheme.
+    #[inline]
```

Comment:
These should either be `inline(always)`, or left to LTO.

Reply:
I don't think this is right,
(1) my experience has been that benchmarks got significantly better in ORAM when I used `#[inline]`, instead of doing what you are suggesting, for functions in `aligned`, `aligned-cmov` and other crates
(2) I think the convention in many essential rust libraries is to use `#[inline]` when we want to permit the optimizer to inline this across crate boundaries, which is usually the case for tiny, low-level functions like this. For example, here's a related discussion in the `hashbrown` crate which is very informative: https://github.com/rust-lang/hashbrown/pull/119

I think it's better to permit the compiler to inline things in cases like this, if llvm decides its a good idea. Thin-LTO is a late stage optimization pass, with less information and less time to make optimization decisions, and it shouldn't be relied on or used as a replacement for things that would normally be done at -O2 or -O3 optimizations.
If we omit `#[inline]` then the optimizer is not permitted to inline this function across crate boundaries, even at `-O2` and `-O3` levels.

The main reason to remove inline annotations is to sacrifice runtime performance for better compile times. We shouldn't do that unless the compile times for this are actually slow. I see no evidence that that is the case.

Reply:
I guess we could have an `inline-more` feature like hashbrown does: https://github.com/rust-lang/hashbrown/blob/master/Cargo.toml#L53

It just seems kind of silly

Reply:
Were you building with lto?